### PR TITLE
Integrate with midje and clojure.test

### PR DESCRIPTION
**Pull request by philomates**
Use matcher-combinators as midje checkers

TODOs:

 - [x] Explore how to implement `embeds`
 - [ ] ~Fix-up printing~ do this in a future PR
 - [x] Explore parsing functions into matchers or pointer comparisons depending if they are midje checkers or not
 - [ ] pick a license

~In order for this to work we need midje to implement these changes: https://github.com/marick/Midje/pull/420~

### CODE REVIEW

File: src/matcher_combinators/core.clj [hunk 1]
```diff
@@ -60,7 +69,7 @@
           selected-value   (select-fn candidate)]
       (select? selected-matcher select-fn selected-value)))
   (match [_this actual]
-    (match-map expected actual identity)))
+    (match-map expected actual identity :contains)))
```
Comment (philomates) [inline src/matcher_combinators/core.clj]: Instead of using a symbol tag to specify the matcher type, we could pass `match-map` the `_this` arg in place of the underlying `expected` value and use things like `(instance? EqualsMap ..)` instead of `(= :equals ..)`. 
Both feel messy in their own regards, so curious if you have a reason to use one or the other.

File: src/matcher_combinators/core.clj [hunk 2]
```diff
@@ -72,20 +81,37 @@
...
 (defrecord EqualsSequence [expected]
   Matcher
   (match [_this actual]
     (if-not (sequential? actual)
       [:mismatch (model/->Mismatch expected actual)]
-      (let [matcher-fns     (concat (map #(partial match %) expected)
-                                    (repeat (fn [extra-element] [:mismatch (model/->Unexpected extra-element)])))
+      ;; TODO PLM: if we want to pass down matcher types between maps/vectors,
+      ;; the `:equals` needs to be dynamically determined
+      (let [matcher-fns     (concat (map #(partial match (derive-matcher % :equals)) expected)
```
Comment (philomates) [inline src/matcher_combinators/core.clj]: the TODO related to https://github.com/rafaeldff/matcher-combinators/pull/1/files#diff-29c6e11ce384b0c0050afc4e3eb7ffd7R23

File: src/matcher_combinators/printer.clj [hunk 3]
```diff
@@ -18,6 +18,9 @@
 (defmethod markup-expression Unexpected [unexpected]
   (list 'unexpected (->ColorTag :red (:actual unexpected))))
 
+(defmethod markup-expression FailedChecker [failed-checker]
+  (list 'checker (->ColorTag :yellow (:form failed-checker)) (->ColorTag :red (:actual failed-checker))))
```
Comment (philomates) [inline src/matcher_combinators/printer.clj]: The `form` information, which shows the checkers name, also shows up in midje, but I figured this lib might not always be use by midje so it might be good to include it.

File: test/matcher_combinators/checkers_test.clj [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  (->Point 1 2) => (ch/match {:x 1 :y 2})
+  {:a (->Point 1 2) :b 2} => (ch/embeds {:a (->Point 1 2)})
+  {:a (->Point 1 2) :b 2} => (ch/embeds {:a {:x 1 :y 2}})
+  (->Point {:a 1} {:b 2}) => (ch/embeds {:x {:a 1}})
+  (->Point {:a [1 2]} {:b 2}) => (ch/embeds {:x {:a (c/in-any-order [2 1])}})
+  (->Point {:a 1 :c 3} {:b 2}) =not=> (ch/embeds {:x {:a 1 :c 6}}))
+
+;; TODO PLM: do we want to allow this?:
+(fact "matching maps with records"
+  {:x 1 :y 2} => (ch/match (->Point 1 2))
```
Comment (philomates) [inline test/matcher_combinators/checkers_test.clj]: Opinions on whether this is good to allow or not?

File: src/matcher_combinators/core.clj [hunk 5]
```diff
@@ -27,116 +37,281 @@
...
+  clojure.lang.IFn
+    (invoke [this actual]
+      (check-match this actual))
+
+  Matcher
+  (match [_this actual]
+    (match-any-order expected actual true)))
+
+;; This is just like midje's `contains :in-any-order :gaps-ok`
+(defn match-subset [expected]
```
Comment (philomates) [inline src/matcher_combinators/core.clj]: `match-subset` and `match-subseq` above aren't good names, but I want to get away from `contains` because I find it confusing. Will try to think of better names

File: src/matcher_combinators/core.clj [hunk 6]
```diff
@@ -1,24 +1,34 @@
 (ns matcher-combinators.core
   (:require [clojure.set :as set]
-            [matcher-combinators.model :as model]))
-
-(defn- find-first [pred coll]
-  (->> coll (filter pred) first))
+            [matcher-combinators.model :as model]
+            [matcher-combinators.printer :as printer]
+            [midje.checking.core :as checking]
```
Comment (rafaeldff) [inline src/matcher_combinators/core.clj]: I wanted to avoid having `core` depend on midje. In theory, this library could be used with clojure test and it's possible to think of non-testing use cases as well.

File: src/matcher_combinators/core.clj [hunk 7]
```diff
@@ -1,24 +1,34 @@
...
 
-(defn- match? [match-result]
+(defn match? [match-result]
   (= :match (first match-result)))
 
 (defn- mismatch? [match-result]
   (= :mismatch (first match-result)))
 
+(defn- check-match [matcher actual]
+  (if (exception/captured-throwable? actual)
```
Comment (rafaeldff) [inline src/matcher_combinators/core.clj]: We can probably make this work with a `defmulti` `check-match` in this namespace and a `defmethod` in `matcher-combinators.midje` defined for whatever is it that `as-data-laden-falsehood` returns.

File: src/matcher_combinators/parser.clj [hunk 8]
```diff
@@ -27,6 +27,3 @@
                BigInteger
                BigInt
                Character)
-
-(mimic-matcher core/equals-map IPersistentMap)
-(mimic-matcher core/equals-sequence IPersistentVector)
```
Comment (rafaeldff) [inline src/matcher_combinators/parser.clj]: I'm curious why did you remove these two lines? They were what enabled the extend-protocol based "parse" to work just like embeds. Is it somehow related to having matchers implement IFn/invoke?

**Review by rafaeldff — COMMENTED** [review]
I'm not sure I'm understanding the direction of this PR, in particular the changes to parser left me scratching my head a bit.

**Comment by rafaeldff** [comment]
The way I was planning on integrating Midje was something like

```clj
(ns my-test
  (:require [matcher-combinators.midje :refer [match]]
            [matcher-combinators.core :refer [equals-map]]))

(fact
   [{:a 2 :b [{:x 42 :y "lala"}] => (match {:a 2 :b [(equals-map {:x 42})]}))
```

Couple of things to note:
1. only `matcher-combinators.midje` refers to midje
2. `matcher-combinators.midje/match` takes a combinator and returns a midje checker 
3. most of the combinators are build implicitly by the "parser" (via extend-protocol), we only specify something like `equals-map` when we want to deviate from the default.

File: src/matcher_combinators/parser.clj (hunk 8 above)
Comment (philomates) [reply src/matcher_combinators/parser.clj]: If I understand correctly the code you had implemented, parsing maps as `equals-map` matchers means that using `embeds-map` will only apply to that one level. Which sort of defeats the purpose, and wouldn't be a drop-in replacement for our internal `embeds` midje checker.

For instance

```clojure
(match 
  (embeds-map
    {:a {:aa {:aaa 1}}})
  {:a {:aa {:aaa 1 :bbb 2} :bb 3} :b 2})
;; => mismatch

;; where as
((internal-impl/embeds
    {:a {:aa {:aaa 1}}})
  {:a {:aa {:aaa 1 :bbb 2} :bb 3} :b 2})
;; => true
```

This explains the `derives-matcher` logic I wrote in `core`

Comment (philomates) [reply src/matcher_combinators/parser.clj]: And there are still subtleties to be figure out with what I'm trying to accomplish in these regards: for instance https://github.com/rafaeldff/matcher-combinators/pull/1/files#r151175091

**Comment by philomates** [comment]
@rafaeldff one thing that I was trying to accomplish is avoid the need to wrap matchers in a `match` in order to use them with midje.

My reasoning is that we use midje checkers a lot, so avoiding the introduction of an additional function call will help keep our tests more readable.

Avoiding `match` is why I made the matchers implement `IFn`, so they can be used as checker functions by midje without wrapping them in anything.

This also introduces some complexity in the matcher code, currently in the form of wrapping core matchers in a `Checker` matcher, and exporting that to be used in midje. This adds an additional `match` call dispatch, but also means that midje tests can import everything from `matcher-combinators.midje` instead of also pulling certain things from `matcher-combinators.core`

File: test/matcher_combinators/checkers_test.clj [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+  {:a [1]} =not=> (c/equals-map {:a (c/in-any-order [(m/as-checker odd?) 1])})
+  [1 2] => (c/in-any-order [(m/as-checker even?)
+                            (m/as-checker odd?)]))
+
+(m/unfinished f)
+(let [short-list (c/equals-sequence [m/anything m/anything m/anything])]
+  (fact "using matchers on the left side of the arrow"
+    (f [1 2 3]) => 1
+    (provided
+      (f short-list) => 1)))
```
Comment (phillipm) [inline test/matcher_combinators/checkers_test.clj]: An additional thing that you get by making matchers into midje checkers (via `as-checker`), is that you can use matchers inside of `provided`. Though you have to bind it to a var, because `provided` doesn't deal well with nested calls like `(f (c/equals-sequence ..)) => 1)`

File: test/matcher_combinators/checkers_test.clj [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+  {:a [1]} =not=> (ch/equals-map {:a (ch/in-any-order [(m/as-checker odd?) 1])})
+  [1 2] => (ch/in-any-order [(m/as-checker even?)
+                            (m/as-checker odd?)]))
+
+(m/unfinished f)
+(let [short-list (ch/equals-sequence [m/anything m/anything m/anything])]
+  (fact "using matchers on the left side of the arrow"
+    (f [1 2 3]) => 1
+    (provided
+      (f short-list) => 1)))
```
Comment (philomates) [inline test/matcher_combinators/checkers_test.clj]: An additional thing that you get by making matchers into midje checkers (via `as-checker`), is that you can use matchers inside of `provided`. Though you have to bind it to a var, because `provided` doesn't deal well with nested calls like `(f (c/equals-sequence ..)) => 1)`

**Comment by rafaeldff** [comment]
I'm not bothered by the requirement of wrapping matchers in a `(match ...)` call because the typical usage (i.e. the embeds-equivalent usage) would not require any other api calls, any map can act as a `contains-map` matcher and any sequence can act as an `equals-sequence` matcher.

```
(ns my-test
  (:require [matcher-combinators.midje :refer [match]]
            [matcher-combinators.core :refer [equals-map]]))

(fact
   [{:a 2 :b [{:x 42}] => (match {:a 2 :b [{:x 42}]}))
```

The above looks pretty similar to how `embeds` is used today, doesn't it? 

If `(match)` returns a midje matcher, then I think it could also be used in `provided` clauses, right?

File: src/matcher_combinators/core.clj [hunk 11]
```diff
@@ -60,9 +103,12 @@
           selected-value   (select-fn candidate)]
       (select? selected-matcher select-fn selected-value)))
   (match [_this actual]
-    (match-map expected actual identity)))
+    (match-map expected actual identity true)))
 
-(defn contains-map [expected]
+(defn contains-map
```
Comment (philomates) [inline src/matcher_combinators/core.clj:86]: how do you feel about renaming this to `embeds`?

**Comment by philomates** [comment]
@rafaeldff Can you take another look at this?

I tried renaming the matchers to allow one to import them using `:refer :all` (not that I encourage this 😸 ) without clashing with core clojure or midje names.
I've also implemented [your suggestion](https://github.com/rafaeldff/matcher-combinators/pull/1#issuecomment-346877110) of wrapping checks in `match` because it is a simpler implementation and can be adapted if needed in the future.
And also included in my new changes is integration with `clojure.test`, so we know that the design approach is compatible with frameworks other than Midje.

What is a good license for this?
Also, seeing that this has potential to be a widely used lib (given that various people have asked for it in Midje issues), would you like to leave this under your personal github or make it a part of nubank's?

**Comment by rafaeldff** [comment]
@phillipm Taking a look.

About the license, Nubank is only approved to use Apache License v2, which is fine by me. Lots of libraries in the Clojure ecosystem adopt EPL, and Midje selected MIT. This is currently on a personal repo with a non specified license because I was still considering it a work in process experiment and didn't want to advertise it widely nor spend a lot of time on these matters at the moment. I'm thinking moving it under nubank's org licensed as APL2 is probably the way to go.

File: src/matcher_combinators/core.clj [hunk 12]
```diff
@@ -27,31 +32,69 @@
...
+
+(defmacro pred->matcher
+  "Turns a normal predicate function into a matcher.
+  Preserves syntactical info when showing mismatch output."
+  [pred & args]
+  (if (empty? args)
+    `(->Predicate ~pred '~pred)
+    `(->Predicate (~pred ~@args) '(~pred ~@args))))
+
+(defn- derive-matcher [matcher-or-pred]
```
Comment (rafaeldff) [inline src/matcher_combinators/core.clj]: Can we get the same behavior extending the Matcher protocol for `IFn`s in `parser` declaring to behave like the `pred->matcher`?

**Review by rafaeldff — COMMENTED** [review]
Commenting on the `derive-matcher` function. I'll continue to review the other parts.

File: src/matcher_combinators/core.clj (hunk 12 above)
Comment (philomates) [reply src/matcher_combinators/core.clj]: I wanted to do this, but it isn't possible in Clojure (as far as I can tell) given that `IFn` isn't a normal protocol: https://stackoverflow.com/a/24975230

Comment (philomates) [reply src/matcher_combinators/core.clj]: Oh, maybe I misread what you are suggesting.. let me try it out

File: src/matcher_combinators/core.clj [hunk 13]
```diff
@@ -1,24 +1,29 @@
...
-
 (defprotocol Matcher
-  (select? [this select-fn candidate])
-  (match [this actual]))
+  "For matching expected and actual values, providing helpful mismatch info on
+  unsucessful matches"
+  (select? [this select-fn candidate]
+           "useful for anchoring specific substructures for `in-any-order` matchers")
+  (match   [this actual]
+           "determine if a concrete `actual` value satisfies this matcher"))
```
Comment (rafaeldff) [inline src/matcher_combinators/core.clj:12]: I'm not sure if we want to make this protocol a public part of the API. It could be an implementation detail initially, to be opened up when we have more confidence the interface is not going to change. (for instance, I'm not 100% happy with the `select?` interface, it feels like it's not on the same semantic level as `match`, but I couldn't think of a better approach).

File: src/matcher_combinators/core.clj [hunk 14]
```diff
@@ -1,24 +1,29 @@
...
 
 (extend-type nil
   Matcher
   (select? [_ _ _] false))
 
-(defn- match? [match-result]
+(defn match? [match-result]
```
Comment (rafaeldff) [inline src/matcher_combinators/core.clj:18]: Why is this function public?

File: src/matcher_combinators/core.clj [hunk 15]
```diff
@@ -1,24 +1,29 @@
...
   (= :match (first match-result)))
 
 (defn- mismatch? [match-result]
   (= :mismatch (first match-result)))
 
+(defn matcher? [x]
```
Comment (rafaeldff) [inline src/matcher_combinators/core.clj:24]: Why is this function public?

File: src/matcher_combinators/core.clj (hunk 11 above)
Comment (rafaeldff) [reply src/matcher_combinators/core.clj:86]: I avoided using the `embeds` name because it's not really the same thing as the nubank-internal existing `embeds`. But thinking more broadly, it might be a good choice. Let me mull it over some more.

File: src/matcher_combinators/core.clj [hunk 16]
```diff
@@ -60,9 +103,12 @@
           selected-value   (select-fn candidate)]
       (select? selected-matcher select-fn selected-value)))
   (match [_this actual]
-    (match-map expected actual identity)))
+    (match-map expected actual identity true)))
 
-(defn contains-map [expected]
+(defn contains-map
+  "Matcher that will match when the map, and any nested maps, contain some of
```
Comment (rafaeldff) [inline src/matcher_combinators/core.clj]: The comment about nested maps is a bit odd. `contains-map` only concerns itself with the map argument. Whatever happens to nested maps will be determined by their matchers. The fact the parser defaults to `contains-map`is unrelated to an enclosing `contains-map` explicit call.

File: src/matcher_combinators/core.clj [hunk 17]
```diff
@@ -72,57 +118,101 @@
           selected-value   (select-fn candidate)]
       (select? selected-matcher select-fn selected-value)))
   (match [_this actual]
-    (match-map expected actual model/->Unexpected)))
+    (match-map expected actual model/->Unexpected false)))
 
-(defn equals-map [expected]
+(defn equals-map
+  "Matcher that will match when the given map is exactly the same as the
+  `expected` map."
```
Comment (rafaeldff) [inline src/matcher_combinators/core.clj]: `Will match when the specified value matchers match exactly all given map values` is more correct, but the wording is awkward.

File: src/matcher_combinators/core.clj [hunk 18]
```diff
@@ -72,57 +118,101 @@
...
+
+(defrecord EqualsSequence [expected]
+  Matcher
+  (match [_this actual]
+    (sequence-match expected actual false)))
 
-(defn equals-sequence [expected]
+(defn equals-seq
+  "Matcher that will match when the given list is exactly the same as the
+  `expected`.
```
Comment (rafaeldff) [inline src/matcher_combinators/core.clj]: `When specified matchers match exactly the given values` is also more correct but awkward.

File: src/matcher_combinators/core.clj [hunk 19]
```diff
@@ -72,57 +118,101 @@
...
+                             #(match? (match (derive-matcher %) first-element))
+                             matchers)]
       (if (nil? matching-matcher)
         false
-        (recur (remove #{matching-matcher} matchers) rest-elements)))))
+        (recur (remove #{matching-matcher} matchers) rest-elements subset?)))))
+
+(defn- match-all-permutations [matchers elements subset?]
+  (helpers/find-first (fn [matchers] (matches-in-any-order? matchers elements subset?))
+              (helpers/permutations matchers)))
```
Comment (rafaeldff) [inline src/matcher_combinators/core.clj:154]: Can you point me to the sample test showing the need for `match-all-permutations`? I had convinced myself something like this wasn't needed, but I'm very often wrong about these things.

File: src/matcher_combinators/core.clj [hunk 20]
```diff
@@ -136,7 +226,42 @@
...
+   (->InAnyOrder expected))
   ([select-fn expected]
    (->SelectingInAnyOrder select-fn expected)))
+
+(defrecord SubSeq [expected]
+  Matcher
+  (match [_this actual]
+    (sequence-match expected actual true)))
+
+(defn sublist
```
Comment (rafaeldff) [inline src/matcher_combinators/core.clj:224]: WDYT of calling this `prefix`?

File: src/matcher_combinators/printer.clj [hunk 21]
```diff
@@ -32,8 +35,11 @@
       (colorized-print markup)
       (pprint/simple-dispatch markup))))
 
-(defn print [expression]
+(defn print [expr]
```
Comment (rafaeldff) [inline src/matcher_combinators/printer.clj]: Aren't we shadowing clojure's `print` here?

File: test/matcher_combinators/checkers_test.clj [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+  {:a {:bb 1} :c 2} => (ch/match {:a {:bb m/anything}})
+  {:a {:bb 1} :c 2} => (ch/match {:a {:bb (m/roughly 1)}}))
+
+(facts "predicates in matchers"
+  (fact "you can use a plain predicate inside matchers"
+    {:a {:bb 1} :c 2} => (ch/match (c/equals-map {:a {:bb odd?} :c 2}))
+    {:a {:bb 1} :c 2} => (ch/match {:a {:bb odd?}}))
+  (fact "but if you want to check exact functions use midje's 'exactly'"
+    {:a {:bb odd?} :c 2} =not=> (ch/match (c/equals-map {:a {:bb odd?} :c 2}))
+    {:a {:bb odd?} :c 2} => (ch/match (c/equals-map {:a {:bb (m/exactly odd?)} :c 2}))))
```
Comment (rafaeldff) [inline test/matcher_combinators/checkers_test.clj]: We probably should provide our own version of `m/exactly` to override the functions-are-predicate-matchers default behavior.

**Review by rafaeldff — COMMENTED** [review]
A few more comments. Overall, I think this PR is on the right track. We should probably merge after this round of Q&A.

File: src/matcher_combinators/core.clj (hunk 19 above)
Comment (philomates) [reply src/matcher_combinators/core.clj:154]: I'm pretty sure https://github.com/rafaeldff/matcher-combinators/pull/1/files#diff-29c6e11ce384b0c0050afc4e3eb7ffd7R93 would fail. The https://github.com/rafaeldff/matcher-combinators/pull/1/commits/ac4c76c101d8811c43b57e3f5eb8d588ee7b1969 commit adds the permutation code and mentions fixing a few then-failing tests

File: src/matcher_combinators/core.clj (hunk 13 above)
Comment (philomates) [reply src/matcher_combinators/core.clj:12]: This would mean people couldn't define their own matcher combinators, which might be useful for some..
Also, how would one make a protocol private? Are you just suggesting we don't document it so that people don't use it?

File: src/matcher_combinators/core.clj (hunk 14 above)
Comment (philomates) [reply src/matcher_combinators/core.clj:18]: because it is used in the `midje` and `test` namespaces. I could lift it into the `helpers` namespace so people don't try to call it.

File: src/matcher_combinators/core.clj (hunk 15 above)
Comment (philomates) [reply src/matcher_combinators/core.clj:24]: same as above, used in other namespaces in the project

File: src/matcher_combinators/core.clj (hunk 20 above)
Comment (philomates) [reply src/matcher_combinators/core.clj:224]: What worries me about `prefix` is that it makes it easily conflated with strings.

File: test/matcher_combinators/checkers_test.clj (hunk 22 above)
Comment (philomates) [reply test/matcher_combinators/checkers_test.clj]: I realized that wrapping the function in `core/equals-value` is equivalent to midje's `exactly`, I created a test example and will mention it when we write docs

File: src/matcher_combinators/parser.clj [hunk 23]
```diff
@@ -11,6 +12,16 @@
...
 
+(extend-type clojure.lang.IFn
+  core/Matcher
+  (match [this actual]
+    (cond
+      (vector? this) (core/match (core/equals-seq this) actual)
+      (map? this)    (core/match (core/contains-map this) actual)
+      :else          (if (this actual)
+                       [:match actual]
+                       [:mismatch (model/->FailedPredicate (str this) actual)]))))
```
Comment (philomates) [inline src/matcher_combinators/parser.clj]: In order to get rid of the `derive-matcher` logic I had to do type dispatching in the `IFn` extension, due to the fact that vectors and maps are also functions, so this `IFn` extension overwrote the default matchers for maps and vectors. I don't really have an opinion as to which one is better, but the changes are all contained in https://github.com/rafaeldff/matcher-combinators/pull/1/commits/64e2702ad767ac2b9cf36baf8b5dc35dc98441f6

**Review by philomates — COMMENTED** [review]
Thanks for the review!
I addressed most of your comments and added a follow-up question about addressing public/private funcs in `core` by moving them to `helpers`

File: src/matcher_combinators/parser.clj (hunk 23 above)
Comment (rafaeldff) [reply src/matcher_combinators/parser.clj]: I thinking extending the protocol to `clojure.lang.Fn` as opposed to my initial suggestion of `clojure.lang.IFn` should allow you to define if for functions but not maps, sets or vectors.

Comment (philomates) [reply src/matcher_combinators/parser.clj]: Ah-ha, that works, thanks!

Comment (rafaeldff) [reply src/matcher_combinators/parser.clj]: Chalk it up one more to Rich Hickey's foresight.

**Review by rafaeldff — APPROVED** [review]
I don't see any more changes required to support midje and clojure.test integration, so I'll go ahead and merge this PR. We can review the finer points of separating the public API (and perhaps an SPI for clients to build their own matchers) and matcher naming in future work.
